File: frontend/lib/models/room.dart
class Room {
  Room({this.occupantId, this.id, required this.name, this.vacant = true, this.occupantName});

  final String? id;
  final String name;
  final bool vacant;
  final String? occupantId;
  final String? occupantName;

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      id: json['_id'],
      name: json['name'],
      vacant: json['vacant'],
      occupantId: json['occupantId'],
      occupantName: json['occupantName'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'vacant': vacant,
      'occupantId': occupantId,
      'occupantName': occupantName,
    };
  }
}

File: backend/models/room.js
import mongoose from "mongoose";

const roomSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    vacant: { type: Boolean, default: true, index: true }, // Indexing for faster vacant room queries
    occupantId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Student",
      default: null,
    }, // Reference for data integrity
    occupantName: { type: String, default: null, trim: true },
  },
  { timestamps: true } // Automatically adds createdAt & updatedAt
);

const Room = mongoose.model("Room", roomSchema);
export default Room;
